docs/signals.txt

```diff
@@ -128,6 +128,7 @@ irc-nicklist.c:
  "server nick changed", SERVER_REC
 
 irc-servers.c:
+ "server outgoing", SERVER_REC, char *str
```

Comment:
This needs a length parameter:

```suggestion
 "server outgoing", SERVER_REC, char *str, int len
```

Reply:
... eww ...

true, but we should avoid char*+len style signal apis, easy to break memory

Reply:
Then we'd need to change the C/Perl boundary to handle a new type that'll do the length param for us.  Perl does have an SvPV(SV *, STRLEN) macro that'll put the length of the string into the second parameter (STRLEN is a macro that usually expands to size_t).

Reply:
Why does this one need it and "server incoming" doesn't? I don't see any other signal passing explicit length arguments after strings

Reply:
Because server incoming is reading from the socket, and that what was read is NULL-terminated by the function doing the reading, so length doesn't matter.  Writing to a socket requires the length to write, which is taken very early and passed around.  Another option could be to change core to do an strlen at the last possible moment, since protocols irssi is likely to ever handle will never allow NULLs, and then this function and corresponding signal don't need a length parameter.

Reply:
Of the two uses of this signal, both present in this PR, one does strlen() explicitly to acquire the length right before calling it, the other takes `gstring->str` which is [guaranteed to be null-terminated][1].

I don't think this function is special, no other code is really prepared to handle strings that aren't null terminated. We should just require the argument to be null terminated for users of the signal, which in practice is already a requirement for every char *

[1]: https://developer.gnome.org/glib/stable/glib-Strings.html

Reply:
..now I realize that you're right in that it needs a length parameter here in signals.txt, because the way it's used in the other parts of the diff passes that length parameter, only so that it can match the existing signature of `irc_server_send_data`. It's all about skipping one strlen for the gstring case, welp.

Reply:
then get rid of the len and good to merge ?

Reply:
Hmm, good point. I didn't think much about this and simply did it the same way as "server incoming" is defined.

@dequis While `gstring->str` is guaranteed to be null-terminated, it can still contain nuls when using `g_string_append_len`.

As I understand it (see the other comment), `irc_server_send_data` is internal API, so changing its signature should not be an issue. For IRC specifically, we could probably get away with dropping the length parameter and only passing the `*char` since IRC does not permit NULs anywhere. So I guess the question is, should "server incoming/outgoing" be generic enough to handle any protocol or IRC-specific (as incoming is currently and outgoing is implemented here)?

Reply:
So, what's the solution here? Remove the length parameter and run `strlen` inside `irc_server_send_data`?

Reply:
@dwfreed any thoughts?